serializations.md

```diff
@@ -1029,3 +1029,539 @@ NOTE:
 , <pubkey>   :: binary()
 ]
 ```
+
+## Serialization of FATE contracts
+
+FATE contracts uses a recursive BNF like definition.
+There are serialisations (RLP, FATE), sequences of serializations (`S1, S2, ...`),
+sequences of bytes (`<< .... >>`) and sequences of bits (`<<< ... >>>`).
+
+Fate Code uses both RLP enodings and pure byte arrays (or binary) encodings.
+In the description here we write `RLP(X)` for the RLP encoding of `X`, where
+X is a name of an encoding described above or below in the description of
+the FATE encoding. We write `<X, Y>` for a byte array made of the sequence of
+the byte encodings of `X` and `Y`. We write `X(Y)` to denote the encoding
+of the type `Y` using the encoding `X`.
+
+We write `< >` for the empty sequence and use `|` for alternatives. We
+use hexadecimal numbers to represent specific bytes used to indicate
+what is encoded, e.g. `0xfe` for a function. We use `_` to denote any
+byte in a byte sequence `<< _ >>` or any encoding in a encoding
+sequence ` _, `.
+
+We use `<<< X, Y >>>` to denote the bitpattern of the bits `X` followed
+by the bits `Y` where the bits on the left (`X`) are the most significant bits.
+A bitpattern should produce a number of bits devisable by 8, if not the pattern
+should be padded with zeroes on the left (in the high order) until it is devisable by 8.
+We use ```<<< xy >>>``` as a shorthand for ```<<< x, y >>>``` when `x` and `y` are
+just 0 or 1 and not a recursive definition.
+
+
+We use `; Text` to denote a comment that is not part of the encoding but
+used to make the meaning of the encoding clear.
+
+### FATE top level
+A fate contract is serialized as follows
+```
+Contract ::=
+  RLP(Code),
+  RLP(Symbols),
+  RLP(Annotations)
+```
+
+### Code
+Functions are sorted on their name (a name in this setting is the 4 first bytes of the blake2b
+hash of the program level readable name). Only one function of each name is allowed.
+```
+Code ::=
+  < >
+| Function , Code
+
+
+```
+
+### Function
+```
+Function ::=
+  0xfe,
+  Id,
+  Attributes,
+  Type Signature,
+  Instructions
+```
+
+### Id
+The Id is four bytes. The convention used by the reference compiler and
+other tools is that this name is the first 4 bytes of the blake2b hash of
+the function name. This is also true for the special `init` function.
+that is called by the `contract_create` transaction. This function
+must have the following Id: `<<0x44, 0xd6, 0x44, 0x1f>>`.
+
+But on a protocol level all other name could be any 4 bytes,
+just as long as all functions in the same contract have different Ids.
+```
+Id ::=
+  << _, _, _, _>>
+
+```
+
+### Attributes
+Attributes are encoded as the data encoding of the integer encoding
+the bit pattern where each bit indicates one attribute as follows:
+* bit 0: private
+* bit 1: payable
+
+```
+Attributes :==
+  Integer(A)
+
+A :==
+  0 ;; No attribute
+  1 ;; private
+  2 ;; payable
+  3 ;; private and payable
+
```

Comment:
reserved to bit 7?

Reply:
These are integer values :-) 0 means no bit set, 1 means bit 1 set, 2 means bit 2 set and 3 means bit 1 and bit 2 set...

Reply:
Perhaps not the best explanation...

Reply:
Now it makes sense. Also it's `Integer(A)` indeed ... 👍